runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -1190,13 +1190,20 @@ public Timer withOutputTimestamp(Instant outputTimestamp) {
      * </ul>
      */
     private void setAndVerifyOutputTimestamp() {
-
       if (outputTimestamp != null) {
-        checkArgument(
-            !outputTimestamp.isBefore(elementInputTimestamp),
-            "output timestamp %s should be after input message timestamp or output timestamp of firing timers %s",
-            outputTimestamp,
-            elementInputTimestamp);
+        // The documentation of getAllowedTimestampSkew explicitly permits Long.MAX_VALUE to be used
+        // for infinite skew. Defend against underflow in that case for timestamps before the epoch.
+        if (fn.getAllowedTimestampSkew().getMillis() != Long.MAX_VALUE
```

Comment:
You can handle the proper bounds via:
```
Instant lowerBound;
try {
   lowerBound = elementInputTimestamp.minus(fn.getAllowedTimestampSkew());
catch (ArithmeticException e) {
   lowerBound = BoundedWindow.TIMESTAMP_MIN_VALUE;
}
if (outputTimestamp.isBefore(lowerBound)) {
  ...
}
```

Finally it would make sense to check the upper bound as well of `BoundedWindow.TIMESTAMP_MAX_VALUE`

Reply:
Sounds good.